packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -2,596 +2,127 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:collection';
+import 'dart:async';
 
-import 'package:flutter/gestures.dart';
-import 'package:flutter/rendering.dart';
 import 'package:flutter/widgets.dart';
-import 'package:quiver/testing/async.dart';
 import 'package:test/test.dart';
 
 import 'binding.dart';
-import 'test_pointer.dart';
+import 'finders.dart';
+import 'controller.dart';
 
-/// Signature for [CommonFinders.byPredicate].
-typedef bool WidgetPredicate(Widget widget);
-
-/// Signature for [CommonFinders.byElement].
-typedef bool ElementPredicate(Element element);
+/// Signature for callback to [testWidgets] and [benchmarkWidgets].
+typedef Future<Null> WidgetTesterCallback(WidgetTester widgetTester);
 
 /// Runs the [callback] inside the Flutter test environment.
 ///
 /// Use this function for testing custom [StatelessWidget]s and
 /// [StatefulWidget]s.
 ///
+/// The callback can be asynchronous (using `async`/`await` or
+/// using explicit [Future]s).
+///
+/// This function is uses the [test] function in the test package to
+/// register the given callback as a test. The callback, when run,
+/// will be given a new instance of [WidgetTester]. The [find] object
+/// provides convenient widget [Finder]s for use with the
+/// [WidgetTester].
+///
 /// Example:
 ///
-///     test('MyWidget', () {
-///        testWidgets((WidgetTester tester) {
-///          tester.pumpWidget(new MyWidget());
-///          tester.tap(find.text('Save'));
-///          expect(tester, hasWidget(find.text('Success')));
-///        });
+///     testWidgets('MyWidget', (WidgetTester tester) {
+///       tester.pumpWidget(new MyWidget());
+///       tester.tap(find.text('Save'));
+///       expect(tester, hasWidget(find.text('Success')));
 ///     });
-void testWidgets(void callback(WidgetTester widgetTester)) {
-  testElementTree((ElementTreeTester elementTreeTester) {
-    callback(new WidgetTester._(elementTreeTester));
-  });
+void testWidgets(String description, WidgetTesterCallback callback) {
+  TestWidgetsFlutterBinding binding = TestWidgetsFlutterBinding.ensureInitialized();
+  WidgetTester tester = new WidgetTester._(binding);
+  group('-', () {
```

Comment:
as discussed in person (but recorded here for posterity), I'm using the group because I need to be able to tear down state when the test times out, and try/finally blocks in the _runTest won't run if the test times out. I've filed https://github.com/dart-lang/test/issues/426 to track this.